lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -0,0 +1,308 @@
+using Org.Apache.REEF.Wake.Remote.Parameters;
+using Org.Apache.REEF.Wake.StreamingCodec;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.Driver
+{
+    /// <summary>
+    /// Implements the IMRU driver on REEF
+    /// </summary>
+    /// <typeparam name="TMapInput">Map Input</typeparam>
+    /// <typeparam name="TMapOutput">Map output</typeparam>
+    /// <typeparam name="TResult">Result</typeparam>
+    internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult> : IObserver<IDriverStarted>,
+        IObserver<IAllocatedEvaluator>, IObserver<IActiveContext>, IObserver<ICompletedTask>
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof (IMRUDriver<TMapInput, TMapOutput, TResult>));
+
+        private readonly ConfigurationManager _configurationManager;
+        private readonly IPartitionedDataSet _dataSet;
+        private readonly IEvaluatorRequestor _evaluatorRequestor;
+        private ICommunicationGroupDriver _commGroup;
+        private readonly IGroupCommDriver _groupCommDriver;
+        private readonly TaskStarter _groupCommTaskStarter;
+        private IConfiguration _tcpPortProviderConfig;
+        private readonly ConcurrentStack<string> _taskIdStack;
+        private readonly ConcurrentStack<IPartitionDescriptor> _partitionDescriptorStack;
+        private readonly int _coresPerMapper;
+        private readonly int _coresForUpdateTask;
+        private readonly int _memoryPerMapper;
+        private readonly int _memoryForUpdateTask;
+        private bool _allocatedUpdateTaskEvaluator;
+        private readonly ConcurrentBag<ICompletedTask> _completedTasks;
+            
+        [Inject]
+        private IMRUDriver(IPartitionedDataSet dataSet,
+            ConfigurationManager configurationManager,
+            IEvaluatorRequestor evaluatorRequestor,
+            [Parameter(typeof (TcpPortRangeStart))] int startingPort,
+            [Parameter(typeof (TcpPortRangeCount))] int portRange,
+            [Parameter(typeof(CoresPerMapper))] int coresPerMapper,
+            [Parameter(typeof(CoresForUpdateTask))] int coresForUpdateTask,
+            [Parameter(typeof(MemoryPerMapper))] int memoryPerMapper,
+            [Parameter(typeof(MemoryForUpdateTask))] int memoryForUpdateTask,
+            IGroupCommDriver groupCommDriver)
+        {
+            _dataSet = dataSet;
+            _configurationManager = configurationManager;
+            _evaluatorRequestor = evaluatorRequestor;
+            _groupCommDriver = groupCommDriver;
+            _coresPerMapper = coresPerMapper;
+            _coresForUpdateTask = coresForUpdateTask;
+            _memoryPerMapper = memoryPerMapper;
+            _memoryForUpdateTask = memoryForUpdateTask;
+            _allocatedUpdateTaskEvaluator = false;
+            _completedTasks = new ConcurrentBag<ICompletedTask>();
+
+            AddGroupCommunicationOperators();
+            ConstructTcpPortProviderConfig(startingPort, portRange);
```

Comment:
What is this needed for? I believe the `TcpPortProvider` configuration is automagically sent to all Evaluators. Do you need to change it here?